src/StoreMeta.cc

```diff
@@ -9,172 +9,147 @@
 /* DEBUG: section 20    Storage Manager Swapfile Metadata */
 
 #include "squid.h"
-#include "base/Range.h"
-#include "MemObject.h"
-#include "Store.h"
+#include "sbuf/Stream.h"
 #include "StoreMeta.h"
-#include "StoreMetaMD5.h"
-#include "StoreMetaObjSize.h"
-#include "StoreMetaSTD.h"
-#include "StoreMetaSTDLFS.h"
-#include "StoreMetaURL.h"
-#include "StoreMetaVary.h"
 
-bool
-StoreMeta::validType(char type)
-{
-    /* VOID is reserved, and new types have to be added as classes */
-    if (type <= STORE_META_VOID || type >= STORE_META_END + 10) {
-        debugs(20, DBG_CRITICAL, "ERROR: storeSwapMetaUnpack: bad type (" << type << ")!");
-        return false;
-    }
-
-    /* Not yet implemented */
-    if (type >= STORE_META_END ||
-            type == STORE_META_STOREURL ||
-            type == STORE_META_VARY_ID) {
-        debugs(20, 3, "storeSwapMetaUnpack: Not yet implemented (" << type << ") in disk metadata");
-        return false;
-    }
+#include <limits>
 
-    /* Unused in any current squid code */
-    if (type == STORE_META_KEY_URL ||
-            type == STORE_META_KEY_SHA ||
-            type == STORE_META_HITMETERING ||
-            type == STORE_META_VALID) {
-        debugs(20, DBG_CRITICAL, "ERROR: Obsolete and unused type (" << type << ") in disk metadata");
-        return false;
-    }
+namespace Store {
 
-    return true;
+/// Whether the given raw swap meta field type represents a type that we should
+/// inform the admin about (if found in a store) but can otherwise ignore.
+inline constexpr
+bool
+DeprecatedSwapMetaType(const RawSwapMetaType type)
+{
+    return
+        type == 1 || // STORE_META_KEY_URL
+        type == 2 || // STORE_META_KEY_SHA
+        type == 6 || // STORE_META_HITMETERING (RFC 2227)
+        type == 7 || // STORE_META_VALID
```

Comment:
@eduard-bagdasaryan, thank you for keeping these deprecated constants from the rest of the code. That was an important design consideration. However, I really do not like the overall result of this change. The duplication of code and casting noise already caused a bug (fixed in ee99367). The earlier simpler PR approach was much better IMO, and I cannot see any good reason to go away from it:

```C++
    return
        type == 1 || // STORE_META_KEY_URL (using URL as the cache key, as in Squid v1.1)
        type == 2 || // STORE_META_KEY_SHA (using SHA for cache key computation instead of MD5)
        type == 6 || // STORE_META_HITMETERING (RFC 2227)
        type == 7; // STORE_META_VALID
```

No duplication. Easy to read/update/review. Fairly safe. All historical information is preserved.

> Please use the enum values by name/label instead of magic numbers here

@yadij, these are not "magic numbers" as explained in another change request comment. There is no harm (and there are benefits) to keeping these numbers unnamed!

Please, let's go back to the earlier style!

Reply:
> Amos: they **are** magic numbers in every sense of the term

Perhaps you have not seen my other [comment](https://github.com/squid-cache/squid/pull/1067#discussion_r897572733) where I have detailed why they are not magic numbers, with references to the definition of the magic number anti-pattern. Applying the (incorrect) suggestion that the PR code was following the magic number auto-pattern would lead to illogical outcomes illustrated by the following [example](https://godbolt.org/z/TxfeYo7so):

```C++
// XXX: A magic number! Terrible code! Fix ASAP!
bool deprecatedType1(const char type)
{
    return type == 6; // STORE_META_HITMETERING (RFC 2227)
}

// No magic here! Great code! Use this pattern everywhere!
bool reservedType2(const char type)
{
    const char STORE_META_HITMETERING = 6; // RFC 2227
    return type == STORE_META_HITMETERING;
}
```

reservedType2() is no better than deprecatedType1() in any area and worse in some areas, but only reservedType2() would pass the (incorrectly applied) "magic numbers" control.

Please do not insist that every singly-used well-documented integer constant gets a name. Doing so makes code worse, including an increased probability of bugs.


> They bypass the whole type validation system of C++ with "magic".

There is no bypass here. Comparing and initializing `char` with small non-negative integers is a safe, common operation. Moreover, the enum-based code uses the very same integers (for readability sake)!

BTW, "magic" in "magic numbers" anti-pattern is not about magical type validation. It is about the unknown (here, "mysterious", "surprising", "magical") origin or meaning of an unnamed undocumented constant. Nothing more. That anti-pattern is very real and quite bad, but the PR code in question does not match that anti-pattern.

Reply:
@rousskov, the anti-pattern that existed in the original PR code was that using a numeric at both enum definition and conditions of if-statements' later prevent the compiler checks and warnings you mention in https://github.com/squid-cache/squid/pull/1067#discussion_r897572733.

Reply:
> The latest iteration of the code meets my requests, so closing this discussion.

I am glad you consider the anti-pattern problem gone. Thank you for resolving this change request.

> the anti-pattern that existed in the original PR code was that using a numeric at both enum definition and conditions of if-statements

FWIW, that pattern never existed -- the original PR code did not use the same numeric values in both places. Here is where numeric values were used for swap meta field types in the original PR commit 0229271 that this change request was attached to:

```
# "enum definition"
0229271:src/StoreMeta.h:    STORE_META_VOID = 0,
0229271:src/StoreMeta.h:    STORE_META_KEY_MD5 = 3,
0229271:src/StoreMeta.h:    STORE_META_URL = 4,
0229271:src/StoreMeta.h:    STORE_META_STD = 5,
0229271:src/StoreMeta.h:    STORE_META_VARY_HEADERS = 8,
0229271:src/StoreMeta.h:    STORE_META_STD_LFS = 9,
0229271:src/StoreMeta.h:    STORE_META_OBJSIZE = 10

# "conditions of if-statements"
0229271:src/StoreMeta.cc:        type == 1 || // STORE_META_KEY_URL
0229271:src/StoreMeta.cc:        type == 2 || // STORE_META_KEY_SHA
0229271:src/StoreMeta.cc:        type == 6 || // STORE_META_HITMETERING (RFC 2227)
0229271:src/StoreMeta.cc:        type == 7 || // STORE_META_VALID
0229271:src/StoreMeta.cc:        type == 11 || // STORE_META_STOREURL (Store-ID URL)
0229271:src/StoreMeta.cc:        type == 12 || // STORE_META_VARY_ID (linking HTTP Vary variants)
```

As you can see, there was no overlap in names and there was no overlap in numeric values. No magic numbers anti-pattern. The code identifying deprecated and reserved types gotten worse while trying to address this unfortunate change request, but those key invariants are still intact.